### ISSUE TITLE
Implement AST RFC

### ISSUE THREAD

**Issue by jazzdan** [facebook/hhvm#5757]
See here: http://3v4l.org/efFoR

Just noting this here to keep track of it.

**Comment by jwatzman** [comment]
This is part of https://wiki.php.net/rfc/abstract_syntax_tree (though probably the most visible part, since HHVM doesn't care about the internal implementation details that comprise of most of that RFC).

**Comment by jwatzman** [comment]
Going to start working on this. https://wiki.php.net/rfc/abstract_syntax_tree#changes_to_syntax_or_behavior is the relevant bits of the RFC.

**Comment by jwatzman** [comment]
Hey @nikic I noticed a discrepancy between what PHP7 does and what the RFC says.

Quoting the RFC:

```
$a = [1, 2];
list($a, $b) = $a;
 
// OLD: $a = 1, $b = 2
// NEW: $a = 1, $b = null + "Undefined index 1"
```

My test case:

```
<?php

error_reporting(-1);

$a = [1, 2];
list($a, $b) = $a;
var_dump($a);
var_dump($b);
```

Output:

```
$ php7 ast.php 
int(1)
int(2)
```

Is this likely to be fixed in PHP7, or should I make HHVM match the behavior?

**Comment by nikic** [comment]
@jwatzman This behavior was adjusted after the RFC as it seemed somewhat unintuitive and people expected doing that to work. Now assignments of that form are special cased to evaluate the RHS first. See http://lxr.php.net/xref/PHP_TRUNK/Zend/zend_compile.c#2565. So yeah, HHVM should implement this as well.

**Comment by jwatzman** [comment]
Cool, will do, thanks.

**Comment by jwatzman** [comment]
Almost done, except this part:

```php
<?php

$obj = new stdClass;
$obj->a = &$obj->b;
$obj->b = 1;
var_dump($obj);
```

This is going to be a PITA/impossible to match with the way HHVM currently compiles this code. There's an ongoing project to split up some of the member instructions, which might help here, so putting on hold until that is ready. Otherwise will just stick this one in inconsistencies and be done with it.

**Comment by jwatzman** [comment]
Nested `list` assignments also need to be done, cf. https://github.com/facebook/hhvm/issues/6049#issuecomment-147004336 -- also waiting on minstr splitting.

**Comment by lexidor** [comment]
I am going over old issues on this repository, to see which ones apply to the current versions of hhvm.

Hhvm's current behavior matches PHP7's.

```
$ hhvm file.php 
int(1)
int(2)
```

```
<?hh // partial

$a = varray[1, 2];
list($a, $b) = $a;
var_dump($a);
var_dump($b);
```

However, do not depend on this being the case forever, see [docs of list()](https://docs.hhvm.com/hack/expressions-and-operators/list).
![image](https://user-images.githubusercontent.com/31805625/82737113-bb9e6d80-9d2e-11ea-83bb-1efdce957739.png)

This warning is based on the earlier doc text:
_If the source elements and the target variables overlap in any way, the behavior is unspecified._
